BLD: Test editable installations for PyWavelets in CI

agriyakhetarpal opened the pull request:
Closes #700. This PR introduces a flag in the Linux matrix setup to run an editable installation of PyWavelets and run the test suite with `pytest`.

I can rebase this branch and revisit this PR after https://github.com/mesonbuild/meson-python/pull/569 is merged and `meson-python==0.16.0` is out on PyPI – it shall fix the failing test as noted in https://github.com/PyWavelets/pywt/issues/700#issuecomment-1959439533.

rgommers commented on pyproject.toml:
```diff
@@ -9,6 +9,7 @@ build-backend = "mesonpy"
 requires = [
     "meson-python>=0.14.0",
     "Cython>=0.29.35",
+    "ninja>=1.10.0",
```
This change is not desirable. `meson-python` will take care of installing `ninja` as necessary

rgommers commented:
Thanks! I'll get your other PR merged first, since there is some overlap. And then you can rebase this on top (CI will also run automatically once you have a PR merged).

agriyakhetarpal commented:
Thanks! I rebased on top of the changes from the previous PR. The failure is expected, of course.

agriyakhetarpal commented:
With `meson-python` v0.16.0 out, this is now ready for review and the test suite against editable installations is working. It has been added for the Linux job for now – we can add it to the macOS job too if needed, but I think testing in just one place is perfectly fine as well. I rebased the branches and have revised the commit history, it should be easier to review/squash.

agriyakhetarpal commented on .github/workflows/tests.yml:
```diff
@@ -100,10 +111,18 @@ jobs:
             fi
 
+          if [ "${OPTIONS_NAME}" == "editable-install" ]; then
+              echo "Testing an editable installation of pywt"
+              sudo apt-get install ninja-build
+              pip install numpy>=2.0.0b1 cython>=3.0.4 meson-python>=0.16.0 matplotlib pytest --force-reinstall
+              pip install -e . --no-build-isolation
```
I had to keep `--force-reinstall` here because it was otherwise using `numpy<2` to compile (which shouldn't be a problem, but good to stay updated since 1.6.0 uses the RC). The workflow matrix is a bit complex so this was the best way that came to mind, I would be happy to explore more ways (probably best left for another PR later on?).

rgommers replied:
`numpy` and `cython` are already installed, so I'd prefer just dropping them here. and then you can also drop `--force-reinstall`. There is no need to reinstall, and I'm fine with just testing with the latest stable release.

agriyakhetarpal commented on pyproject.toml:
```diff
@@ -7,7 +7,7 @@
 build-backend = "mesonpy"
 requires = [
-    "meson-python>=0.15.0",
+    "meson-python>=0.16.0",
```
I bumped the version so that users won't face an issue and that those who are using a development edition of PyWavelets will receive a new version of `meson-python` the next time they retrieve the latest changes. What I'm not sure about is if we should do an exact pin to `0.16.0` though or if the updated lower bound is fine, i.e., in case a newer version of `meson-python` breaks the installations?

agriyakhetarpal replied:
Thiugh if it does, we can always pin it later, so I would lean towards not pinning since most users will be using the wheels anyway.

rgommers replied:
We should never add an upper bound in `main`. In releases we do add an upper bound to dependencies that could get breaking updates, for exactly the reason you mention.

agriyakhetarpal commented:
From the workflow run: https://github.com/PyWavelets/pywt/actions/runs/8723177022/job/23930800559?pr=702#step:4:446

this line shows that the wheel being built is an editable one, but the wheel does not have the `editable` keyword after the version and before the ABI tag – and I think this is a bug, it is not conformant with [PEP 660](https://peps.python.org/pep-0660/). This is not the case with `setuptools` and other build-backends (just verified with both a Hatch-based pure Python project and a `setuptools` + `pybind11` compiled one).

agriyakhetarpal commented:
From the workflow run: https://github.com/PyWavelets/pywt/actions/runs/8723177022/job/23930800559?pr=702#step:4:446

this line shows that the wheel being built is an editable one, but the wheel does not have the `editable` keyword after the version and before the ABI tag – and I think this is a bug, it is not conformant with [PEP 660](https://peps.python.org/pep-0660/). This is not the case with `setuptools` and other build-backends (just verified with both a Hatch-based pure Python project and a `setuptools` + `pybind11` compiled one). I'll file this on the `meson-python` issue tracker.

rgommers commented:
> It has been added for the Linux job for now – we can add it to the macOS job too if needed, but I think testing in just one place is perfectly fine as well.

Agreed, a single job is fine.

rgommers reviewed (commented):
Looks pretty good, only two minor comments.

agriyakhetarpal commented:
Thank you for the review, @rgommers! I have addressed all requested changes and rebased (but only because they were minor). Let's hope that CI passes as usual.

rgommers reviewed (approved):
LGTM now, in it goes. Thanks @agriyakhetarpal, nice work!
